### PR TITLE
CA-199734: best-effort DP destroy and other cleanups for mirror/copy task failures

### PR DESCRIPTION
**Pull request by zli**
This pull req does best-effort DP destroy and other cleanup actions in case of mirror/copy task failures. Along the way, we also made small amount of refactoring to  make the code cleaner and more concise. See the summary of each commit for details.

Signed-off-by: Zheng Li <dev@zheng.li>

### CODE REVIEW

**Comment by jonludlam** [comment]
@zli Please make a better title and description, as these get into the commit history.

**Comment by jonludlam** [comment]
This is certainly a big improvement over the original. However, there's still a window for problems, I think - if an exception happens here: https://github.com/zli/xen-api/blob/a24e501f867f0f05d2b7bec8bb53d06ac6451baf/ocaml/xapi/xapi_vm_migrate.ml#L541-L568
we'll still leak, no?

**Comment by zli** [comment]
> However, there's still a window for problems...

Indeed. I though that part of code is pretty safe as the main activity (copy/mirror) must have been completed by then. But yes, there is still chance the remote query might fail. I can rework a little bit to have the try..with cover that section.

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 1]
```diff
@@ -499,40 +499,45 @@ let vdi_copy_fun __context dbg vdi_map remote is_intra_pool remote_vdis so_far t
...
         end
       in
-      
+
+      let mapfn =
+        let start = (Int64.to_float !so_far) /. (Int64.to_float total_size) in
+        let len = (Int64.to_float vconf.size) /. (Int64.to_float total_size) in
+        fun x -> start +. x *. len
+      in
+
```
Comment (lindig) [inline ocaml/xapi/xapi_vm_migrate.ml]: Minor suggestion to make this code clearer. The name `so_far` suggests that we have something done and need to do more. I also think it is clearer to expose the argument `x` early. I'd write:


    let mapfn x =
      let total     = Int64.to_float total_size in
      let done_     = Int64.to_float !so_far ./ total in
      let remaining = Int64.to_float vconf.size /. total in
        done_ +. x *. remaining

Comment (zli) [inline ocaml/xapi/xapi_vm_migrate.ml]: Integrated as a separate commit, thanks!

**Comment by zli** [comment]
@jonludlam , it's now updated.

- Add pull req title & description
- Changed the exception handling to cover larger range, also did some surgery with the code organization
- More code refactoring as suggested by @lindig

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 2]
```diff
@@ -493,12 +493,40 @@ let vdi_copy_fun __context dbg vdi_map remote is_intra_pool remote_vdis so_far t
...
+    (* And destroy the new VDI *)
+    if !remote_vdi_reference <> Ref.null then (try XenAPI.VDI.destroy remote.rpc remote.session !remote_vdi_reference with _ -> error "Failed to destroy remote VDI");
+    if mirror_failed then raise (Api_errors.Server_error(Api_errors.mirror_failed,[Ref.string_of vconf.vdi])) else raise e in
+
+  try
     if not mirror then
       let dest_vdi_ref = XenAPI.VDI.get_by_uuid remote.rpc remote.session vdi_uuid in
-      vconf.location,dest_vdi_ref,None,None
+      remote_vdi := vconf.location;
+      remote_vdi_reference := dest_vdi_ref
```
Comment (jonludlam) [inline ocaml/xapi/xapi_vm_migrate.ml]: This will causes us to try to destroy a remote VDI that wasn't made by us in the exception handler, wont it?

Comment (zli) [inline ocaml/xapi/xapi_vm_migrate.ml]: Well spot! The exception handling logic should be aware if the remote disk wasn't created by mirror/copy and action accordingly.  Code is updated.

**Comment by jonludlam** [comment]
Hi @zli. I've been having some philosophical discussions with @simonjbeaumont and @johnelse about this PR. There has been a fair bit of refactoring recently to move to a continuation-passing style implementation in this module. Although, as you say, the logic of the cleanup seems quite reasonable and clear, we've been wondering how it would look when it's refactored into a 'with_dp', 'with_remote_vdi' or some similar formulation. Essentially having smaller functions that do more specific things with their own smaller exception handlers all chained together. I'd really like to see this in comparison with your current more imperative implementation. We can keep this pull request open and then make a decision about which one is clearer, as, although I think it likely that the CPS style will end up being simpler, I'm honestly not sure at this point whether it will be better or not.

I'm quite keen to hear other peoples views on this too.

**Comment by jonludlam** [comment]
I've marked this as blocked until we can compare with the CPS style fix.

**Comment by jonludlam** [comment]
@zli - please prepare an alternative PR as described above. Thanks!

**Comment by zli** [comment]
@jonludlam: sure, will create a separate one.

**Comment by jonludlam** [comment]
Closed in favour of the updated PR

**Pull request closed without merging**